### PR TITLE
Credits integration

### PR DESCRIPTION
**Pull request by rkerins91**
Notes on this PR: 

- I created the functionality for adding credits to user, and temporarily disabled the credit card form because it was non functional and the front end ui had no way to add credits until it is.
- I also connected credits to Alex's code upload component so that when the user attempts to upload code, if the user has credits a credit is removed and the code upload is successful, if the user does not have credits in account, for now there is an alert that informs them of that.
-Also updated my route to add credits to return a 403 (not sure if that's the right error code) if the user is attempting to spend one and does not have enough.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,7 @@ client/.env
 server/.env
 
 client/.env
+
+client/src/pages/PurchaseCredits/CardForm.js
+client/src/pages/PurchaseCredits/CardSection.js
+client/package-lock.json
```
Comment (ashan36) [inline .gitignore]: Shouldn't ignore package-lock, not sure why these other files are being ignored either.

**Review by ashan36 — CHANGES_REQUESTED** [review]
Noticed this branch is out of sync with dev. We should probably be updating feature branches with changes from dev before doing a PR.

File: .gitignore (hunk 1 above)
Comment (rkerins91) [reply .gitignore]: Yeah all this was temporary. The package-lock had a lot of merge conflicts and i wasn't sure which changes to accept. The other two files may or may not be used when use Material UI with react-stripe-elements, so I figured it was better to not push it and have you guys look at possibly obsolete code.

**Comment by rkerins91** [comment]
> Noticed this branch is out of sync with dev. We should probably be updating feature branches with changes from dev before doing a PR.

Thats strange, I just pulled dev and merged it in right before I made this PR. What about it is out of sync?

File: server/database/models/User.js [hunk 2]
```diff
@@ -4,28 +4,33 @@ const jwt = require("jsonwebtoken");
...
+    required: true
+  },
+  experience: {
+    type: Schema.Types.Mixed,
+    of: Number
+  },
+  assigned_threads: [{ type: Schema.Types.ObjectId, ref: Thread }],
+  registeredDate: {
+    type: Date,
+    default: Date.now
```
Comment (ashan36) [inline server/database/models/User.js]: registeredDate is no longer a path on User as that's being store in the createdAt path which is automatically created with the timestamps: true option.

File: client/src/pages/PurchaseCredits/PurchaseCredits.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const incrementCredits = () => {
+    setCredits(++credits);
+  };
+
+  // WILL USE THIS CODE ONCE I ADD CODE FOR CHECKOUT
+  // const handleToggle = () => {
+  //   setCheckout(!checkout);
+  // };
+
+  const handleSubmit = async () => {
```
Comment (ashan36) [inline client/src/pages/PurchaseCredits/PurchaseCredits.js]: Allows user to purchase 0 credits.

File: client/src/pages/PurchaseCredits/PurchaseCredits.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import React, { useState, useContext } from "react";
+import { StripeProvider, Elements } from "react-stripe-elements";
+// import { Grid, Typography, Button, makeStyles, Paper } from "@material-ui/core";
+import { UserContext } from "context/UserContext";
+import axios from "axios";
+import NameForm from "./NameForm";
+// import Cart from "./Cart";
+import SelectAmount from "./SelectAmount";
+// import CheckoutForm from "./CheckoutForm";
+import CardSection from "./CardSection";
```
Comment (ashan36) [inline client/src/pages/PurchaseCredits/PurchaseCredits.js]: Had to comment this out to test after compilation error. Looks like one of the gitignored files

File: client/src/pages/CodeUpload.js [hunk 5]
```diff
@@ -128,13 +128,19 @@ const CodeUpload = () => {
         data: JSON.stringify(requestData)
       });
       //redirect user to their reviews page
-      if (response.data.success) {
+      const removeCredit = await axios.put(`/user/${user._id}/add-credit`, {
+        credits: -1
+      });
+      if (response.data.success && removeCredit.data.success) {
         var alerts = new Set();
```
Comment (ashan36) [inline client/src/pages/CodeUpload.js]: Should probably attempt to remove credits and only after that's successful, post the request.

File: server/database/models/User.js (hunk 2 above)
Comment (rkerins91) [reply server/database/models/User.js]: I fixed this, it must have accepted the wrong changes when resolving conflicts locally. Did you notice anything else out of sync?

File: client/src/pages/PurchaseCredits/PurchaseCredits.js (hunk 3 above)
Comment (rkerins91) [reply client/src/pages/PurchaseCredits/PurchaseCredits.js]: Fixed

File: client/src/pages/CodeUpload.js (hunk 5 above)
Comment (rkerins91) [reply client/src/pages/CodeUpload.js]: Fixed, also added error snackbar when user doesn't have enough credits.

File: client/src/pages/PurchaseCredits/PurchaseCredits.js (hunk 4 above)
Comment (rkerins91) [reply client/src/pages/PurchaseCredits/PurchaseCredits.js]: I commented it out for now. Will stop git ignoring in next PR when those components are ready.

**Pull request merged**